apps/emqx/src/emqx_broker_mon.erl

```diff
@@ -178,6 +208,203 @@ update_broker_pool_max_mailbox_size(State) ->
     atomics:put(Atomics, Idx, MailboxSize),
     MailboxSize.
 
+update_top_n_mailbox_procs(ProcsInfo) ->
+    ets:delete_all_objects(?MAILBOX_LEN_TAB),
+    lists:foreach(
+        fun({Pid, QLen, ProcName}) ->
+            ets:insert(?MAILBOX_LEN_TAB, {Pid, ProcName, QLen, make_proc_meta_data(Pid)})
+        end,
+        ProcsInfo
+    ).
+
+make_proc_meta_data(Pid) ->
+    case erlang:process_info(Pid, dictionary) of
+        undefined ->
+            #{};
+        {dictionary, Dict} ->
+            LoggerProcMeta = proplists:get_value('$logger_metadata$', Dict, #{}),
+            emqx_utils_maps:jsonable_map(
+                LoggerProcMeta,
+                fun(K, V) ->
+                    {emqx_utils_conv:bin(K), emqx_utils_conv:bin(V)}
+                end
+            )
+    end.
+
+non_zero_mailbox_procs(TopN) ->
+    [
+        {Pid, QLen, proc_name(Pid)}
+     || {Pid, QLen, _} <- recon:proc_count(message_queue_len, TopN), QLen > 0
+    ].
+
+make_top_n_mailbox_alarm_msg([]) ->
+    <<"some processes are overloaded">>;
+make_top_n_mailbox_alarm_msg(ProcsInfo) ->
+    [
+        <<"some processes are overloaded, overloaded procs:">>,
+        [
+            [<<" {">>, ProcName, <<": ">>, integer_to_binary(QLen), <<"}">>]
+         || {_Pid, QLen, ProcName} <- ProcsInfo
+        ]
+    ].
+
+make_top_n_mailbox_alarm_detail([]) ->
+    [];
+make_top_n_mailbox_alarm_detail([{_Pid, QLen, _} | _] = ProcsInfo) ->
+    AlarmTopN = overall_mailbox_alarm_top_n(),
+    case QLen >= overall_mailbox_alarm_threshold() of
+        true ->
+            ProcsToBeAlarm = [Pid || {Pid, _, _} <- lists:sublist(ProcsInfo, 1, AlarmTopN)],
+            format_procs_info(ProcsToBeAlarm, verbose);
+        false ->
+            []
+    end.
+
+format_procs_info(PidL, Type) ->
+    lists:filtermap(
+        fun(Pid) ->
+            case format_proc_info(get_pid(Pid), Type) of
+                undefined -> false;
+                Info -> {true, Info}
+            end
+        end,
+        PidL
+    ).
+
+basic_proc_info(Pid, Name, Init, Dict, QLen, Verbose) ->
+    Info0 = #{
+        pid => list_to_binary(pid_to_list(Pid)),
+        registered_name => reg_name(Name),
+        initial_call => init_call(Init, Dict),
+        label => get_label(Pid)
+    },
+    maybe_peek_messages(Pid, Info0, QLen, Dict, Verbose).
+
+maybe_peek_messages(Pid, Info0, QLen, Dict, verbose) when QLen > 0, QLen < 1000 ->
+    Info = Info0#{message_queue_len => QLen, proc_dict_len => length(Dict)},
+    case erlang:process_info(Pid, messages) of
+        {messages, Messages} ->
```

Comment:
My 2 cents: again, this can make things actually worse when there's actual overload. This implies copying the whole mailbox, which can contain 100s of thousands of messages, and also implies costly inter-process synchronization.

Reply:
Here it will get the content of mailbox only when `QLen > 0, QLen < 100`, will this sill have the problem?